packages/loopring_v3/contracts/aux/fast-withdrawal/FastWithdrawalLiquidityProvider.sol

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../agents/FastWithdrawalAgent.sol";
+import "../../lib/OwnerManagable.sol";
+import "../../lib/EIP712.sol";
+import "../../lib/ERC20.sol";
+import "../../lib/ERC20SafeTransfer.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/ReentrancyGuard.sol";
+import "../../lib/SignatureUtil.sol";
+
+
+/// @title Basic contract storing funds for a liquidity provider.
+/// @author Brecht Devos - <brecht@loopring.org>
+contract FastWithdrawalLiquidityProvider is ReentrancyGuard, OwnerManagable
+{
+    using AddressUtil       for address;
+    using AddressUtil       for address payable;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using SignatureUtil     for bytes32;
+
+    struct FastWithdrawalApproval
+    {
+        address exchange;
+        address from;
+        address to;
+        address token;
+        uint96  amount;
+        uint32  storageID;
+        uint64  validUntil; // most significant 32 bits as block height, least significant 32 bits as block time
+        address signer;
+        bytes   signature;  // signer's signature
+    }
+
+    bytes32 constant public FASTWITHDRAWAL_APPROVAL_TYPEHASH = keccak256(
+        "FastWithdrawalApproval(address exchange,address from,address to,address token,uint96 amount,uint32 storageID,uint64 validUntil)"
+    );
+
+    bytes32 public DOMAIN_SEPARATOR;
+
+    FastWithdrawalAgent public agent;
+
+    constructor(FastWithdrawalAgent _agent)
+    {
+        agent = _agent;
+        DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("FastWithdrawalLiquidityProvider", "1.0", address(this)));
+    }
+
+    // Execute the fast withdrawals.
+    // Full approvals are posted onchain so they can be used by anyone to speed up the
+    // withdrawal by calling , but the approvals are not validated when done by the
+    // owner or one of the managers.
+    function execute(FastWithdrawalApproval[] calldata approvals)
+        external
+        nonReentrant
+    {
+        // Prepare the data and validate the approvals when necessary
+        FastWithdrawalAgent.Withdrawal[] memory withdrawals = new FastWithdrawalAgent.Withdrawal[](approvals.length);
+        for (uint i = 0; i < approvals.length; i++) {
+            if (!isManager(msg.sender)) {
```

Comment:
This'll do an SLOAD at 800 gas/withdrawal instead of a single SLOAD for all withdrawals. Not a huge difference but caching it is easy enough.

Reply:
Good point, I'll restore it as you have implemented.

Reply:
After a second thought, I want to make sure every approval is valid (with the right signature) so users wont' worry his own speeding-up the pending txs will be a waste of his own gas.

Reply:
Oops missed this. But users can easily check if the signature is valid before even doing the transaction themselves, no?